docs/04-reference/winglang-spec.md

```diff
@@ -869,30 +869,223 @@ let f = (arg1: num, var arg2: num) => {
+* The `x?.y` notation can be used to access object fields only if there is a value. It returns an undefined optional if the value is not defined.
+* The `x ?? y` notation will return the value in `x` (if there is one) or `y` otherwise (`y` must be of type `T`, not `T?`).
+* A default value can also be supplied using `= y` in the declaration of the field or argument, which implies that the type will always be defined.
 
-<details><summary>Equivalent TypeScript Code</summary>
+#### 1.7.1 Declaration
 
-> ```TS
-> const x: number? = 44;
-> const y: number = x ?? 55;
-> ```
+##### 1.7.1.1 Declaration in struct fields
 
-</details>
+```js
+struct Person {
+  name: str;
+  address: str?;
+}
+```
+
+In the above struct, the `address` field, the `foo` argument and the `my_opt` field are all marked
+as optionals using `?`.
+
+This means, for example, that a `Person` can be initialized without defining the `address`
+field:
+
+```js
+let my_person = Person { name: "david" };
+assert(my_person.address? == false);
+```
+
+It is also possible to use `=` in the field declaration to indicate the default value (similar to function arguments):
+
+```js
+struct FormatOpts {
+  radix: num = 10;
+}
+
+let opts = FormatOpts {};
+assert(opts.radix == 10);
+```
+
+##### 1.7.1.2 Declaration in variables
+
+Use `T?` to indicate that a variable is optional:
+
+```js
+let var y: num?;
+assert(y? == false); // y doesn't have a value
+
+// ok to reassign another value because `y` is reassignable (`var`)
+y = 123;
+assert(y? == true);
+```
+
+It is okay to initialize a value:
+
+```js
+let var y: num? = 12;
+```
+
+##### 1.7.1.3 Declaration in class fields
+
+Similarly to struct fields, fields of classes can be also defined as optional using `T?`:
+
+```js
+class Foo {
+  my_opt: num?;
+  var my_var: str?;
+
+  init(opt: num?) {
+    this.my_opt = opt;
+  }
+
+  set_my_var(x: str) {
+    this.my_var = x;
+  }
+}
+```
+
+##### 1.7.1.4 Declaration in function arguments
+
+In the following example, the argument `by` is optional, so it is possible to call `x()` without
+supplying a value for `by`:
+
+```js
+let increment = (x: num, by: num?): num => {
+  return x + (by ?? 1);
+};
+
+assert(my_function(88) == 89);
+assert(my_function(88, 2) == 90);
+```
+
+Alternatively, specifying a default value will also make an argument optional. Using a default value
+in the function declaration ensures that `by` always has a value so there is no need to unwrap it
+(this is why its type is `num` and not `num?`):
+
+```js
+let increment = (x: num, by: num = 1): num {
+  return x + by;
+}
+```
+
+Non-optional arguments can only be used before all optional arguments:
+
+```js
+let my_fun = (a: str, x?: num, y: str): void = { /* ... */ };
+//-----------------------------^^^^^^ ERROR: cannot declare a non-optional argument after an optional
+```
+
+If a function uses a keyword argument struct as the last argument, and there are other optional arguments before,
```

Comment:
(nit) curious what "keyword argument" means here

Reply:
It's when you can assign the fields of a struct through keyword arguments. Added another example.